copilot/cmd/sidecar.go

```diff
@@ -0,0 +1,186 @@
+	}
+	outputInterface := iface.Outputs
+
+	if iface.Outputs == nil || iface.Outputs.Variables == nil || len(iface.Outputs.Variables) == 0 {
+		logger.Infof(ctx, "Empty output interface received. Assuming void outputs. Sidecar will exit immediately.")
+		return nil
+	}
+
+	f, ok := core.DataLoadingConfig_LiteralMapFormat_value[u.metadataFormat]
+	if !ok {
+		return fmt.Errorf("incorrect input data format specified, given [%s], possible values [%+v]", u.metadataFormat, GetFormatVals())
+	}
+
+	m, ok := core.IOStrategy_UploadMode_value[u.uploadMode]
+	if !ok {
+		return fmt.Errorf("incorrect input upload mode specified, given [%s], possible values [%+v]", u.uploadMode, GetUploadModeVals())
+	}
+
+	logger.Infof(ctx, "Creating start watcher type: %s", u.startWatcherType)
+	w, err := u.createWatcher(ctx, u.startWatcherType)
+	if err != nil {
+		return err
+	}
+
+	logger.Infof(ctx, "Waiting for Container to start with timeout %s.", u.containerStartTimeout)
+	childCtx, cancelFn := context.WithTimeout(ctx, u.containerStartTimeout)
+	defer cancelFn()
+	err = w.WaitToStart(childCtx)
+	if err != nil && err != containerwatcher.ErrTimeout {
+		return err
+	}
+
+	if err != nil {
+		logger.Warnf(ctx, "Container start detection aborted, :%s", err.Error())
+	}
+
+	if u.startWatcherType != u.exitWatcherType {
+		logger.Infof(ctx, "Creating watcher type: %s", u.exitWatcherType)
+		w, err = u.createWatcher(ctx, u.exitWatcherType)
+		if err != nil {
+			return err
+		}
+	}
+
+	logger.Infof(ctx, "Waiting for Container to exit.")
+	if err := w.WaitToExit(ctx); err != nil {
+		logger.Errorf(ctx, "Failed waiting for container to exit. Err: %s", err)
+		return err
+	}
+
+	logger.Infof(ctx, "Container Exited! uploading data.")
+
+	// TODO maybe we should just take the meta output path as an input argument
+	toOutputPath, err := u.Store.ConstructReference(ctx, storage.DataReference(u.remoteOutputsPrefix), u.metaOutputName)
+	if err != nil {
+		return err
+	}
+
+	dl := data.NewUploader(ctx, u.Store, core.DataLoadingConfig_LiteralMapFormat(f), core.IOStrategy_UploadMode(m), ErrorFile)
+
+	childCtx, cancelFn = context.WithTimeout(ctx, u.timeout)
+	defer cancelFn()
+	if err := dl.RecursiveUpload(childCtx, outputInterface, u.localDirectoryPath, toOutputPath, storage.DataReference(u.remoteOutputsRawPrefix)); err != nil {
+		logger.Errorf(ctx, "Uploading failed, err %s", err)
+		return err
+	}
+
+	logger.Infof(ctx, "Uploader completed successfully!")
+	return nil
+}
+
+func (u *UploadOptions) Sidecar(ctx context.Context) error {
+
+	if err := u.uploader(ctx); err != nil {
+		logger.Errorf(ctx, "Uploading failed, err %s", err)
+		if err := u.UploadError(ctx, "OutputUploadFailed", err, storage.DataReference(u.remoteOutputsPrefix)); err != nil {
+			logger.Errorf(ctx, "Failed to write error document, err :%s", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func NewUploadCommand(opts *RootOptions) *cobra.Command {
```

Comment:
In general, I feel like the uploader, downloader, and sidecar watcher should be different cmds in flytectl.  That could get messy too but maybe worth thinking about?

Reply:
no this is not messy at all. I think I agree, I think they should be in flytectl, just that we need to move some code to flyteidl before we do that. What do you think we move ahead and then clean up after the fact. I can drop all the github workflows?

Reply:
sure sure, sounds good.